gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-common/src/main/java/io/gravitee/am/gateway/handler/common/auth/user/impl/UserAuthenticationServiceImpl.java

```diff
@@ -184,10 +186,29 @@ private boolean isIndefinitelyLocked(User user) {
     private Completable checkAccountStatus(User user) {
         if (!user.isEnabled()) {
             return Completable.error(new AccountDisabledException("Account is disabled for user " + user.getUsername()));
+        } else if (checkAccountPasswordExpiry(user)) {
+            return Completable.error( new AccountPasswordExpiredException("Account's password is expired "));
         }
         return Completable.complete();
     }
 
+    /**
+     * Check the user password status
+     * @param user Authenticated user
+     * @return True if the password has expired or False if not
+     */
+    private boolean checkAccountPasswordExpiry(User user) {
+        /** If the expiryDate is set to 0 so it's disabled */
+        if(domain.getPasswordSettings() == null || (domain.getPasswordSettings() != null && domain.getPasswordSettings().getExpiryDuration() == 0))
+            return false;
+
+        Calendar calendar = Calendar.getInstance();
+        calendar.setTime(user.getLastPasswordReset());
```

Comment:
the lastPasswordReset is null for users that exist before the 3.16 release.
When the user is read from the database, maybe we should define the lastPasswordReset to user creation date to avoid this. You did it into the UserSevice but we have to do it also here (or at the Repository level)